File: src/testpki.rs
/*
    Copyright 2023 Loophole Labs

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

           http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

#[derive(Debug, Clone)]
pub struct TestPki {
    pub ca_cert: Vec<u8>,
    pub client_cert: Vec<u8>,
    pub client_key: Vec<u8>,
    pub server_cert: Vec<u8>,
    pub server_key: Vec<u8>
}

impl TestPki {
    #[allow(dead_code)]
    pub(crate) fn new() -> Self {
        let alg = &rcgen::PKCS_ECDSA_P256_SHA256;
        let mut ca_params = rcgen::CertificateParams::new(Vec::new());
        ca_params
            .distinguished_name
            .push(rcgen::DnType::OrganizationName, "TestPKI Server Acceptor");
        ca_params
            .distinguished_name
            .push(rcgen::DnType::CommonName, "TestPKI CA");
        ca_params.is_ca = rcgen::IsCa::Ca(rcgen::BasicConstraints::Unconstrained);
        ca_params.key_usages = vec![
            rcgen::KeyUsagePurpose::KeyCertSign,
            rcgen::KeyUsagePurpose::DigitalSignature,
            rcgen::KeyUsagePurpose::CrlSign,
        ];
        ca_params.alg = alg;
        let ca_cert = rcgen::Certificate::from_params(ca_params).unwrap();

        let mut server_params = rcgen::CertificateParams::new(vec!["localhost".to_string()]);
        server_params.is_ca = rcgen::IsCa::NoCa;
        server_params.extended_key_usages = vec![rcgen::ExtendedKeyUsagePurpose::ServerAuth];
        server_params.alg = alg;
        let server_cert = rcgen::Certificate::from_params(server_params).unwrap();

        let mut client_params = rcgen::CertificateParams::new(Vec::new());
        client_params
            .distinguished_name
            .push(rcgen::DnType::CommonName, "TestPKI Client");
        client_params.is_ca = rcgen::IsCa::NoCa;
        client_params.extended_key_usages = vec![rcgen::ExtendedKeyUsagePurpose::ClientAuth];
        client_params.alg = alg;
        client_params.serial_number = Some(rcgen::SerialNumber::from(vec![0xC0, 0xFF, 0xEE]));
        let client_cert = rcgen::Certificate::from_params(client_params).unwrap();

        Self {
            ca_cert: Vec::from(ca_cert.serialize_pem().unwrap().as_bytes()),
            client_cert: Vec::from(client_cert.serialize_pem_with_signer(&ca_cert).unwrap().as_bytes()),
            client_key: Vec::from(client_cert.serialize_private_key_pem().as_bytes()),
            server_cert: Vec::from(server_cert.serialize_pem_with_signer(&ca_cert).unwrap().as_bytes()),
            server_key: Vec::from(server_cert.serialize_private_key_pem().as_bytes())
        }
    }
}

#[cfg(test)]
mod tests {
    use rustls::client::danger::ServerCertVerifier;
    use rustls::client::WebPkiServerVerifier;
    use rustls::server::WebPkiClientVerifier;
    use rustls_pki_types::UnixTime;
    use super::*;
    use crate::config;

    #[test]
    fn test_server_cert() {
        let test_pki = TestPki::new();
        let root = config::load_ca(&test_pki.ca_cert).unwrap();
        assert_eq!(root.len(), 1);

        let verifier = WebPkiServerVerifier::builder(root.into()).build().unwrap();

        let server_cert = config::load_certs(&test_pki.server_cert).unwrap();
        let (server_cert, intermediate) = server_cert.split_first().unwrap();
        let server_name = "localhost".try_into().unwrap();
        verifier.verify_server_cert(server_cert, intermediate,  &server_name, &[], UnixTime::now()).unwrap();
    }

    #[test]
    fn test_client_cert() {
        let test_pki = TestPki::new();
        let root = config::load_ca(&test_pki.ca_cert).unwrap();
        assert_eq!(root.len(), 1);

        let verifier = WebPkiClientVerifier::builder(root.into()).build().unwrap();

        let client_cert = config::load_certs(&test_pki.client_cert).unwrap();
        let (client_cert, intermediate) = client_cert.split_first().unwrap();
        verifier.verify_client_cert(client_cert, intermediate, UnixTime::now()).unwrap();
    }
}

File: internal/testpki/testpki.go
/*
   Copyright 2023 Loophole Labs

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

          http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

package testpki

import (
	"bytes"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/tls"
	"crypto/x509"
	"crypto/x509/pkix"
	"encoding/pem"
	"fmt"
	"math/big"
	"time"
)

type TestPKI struct {
	CaCert       []byte
	ClientCert   []byte
	ClientKey    []byte
	ClientConfig *tls.Config
	ServerCert   []byte
	ServerKey    []byte
	ServerConfig *tls.Config
}

func New() (*TestPKI, error) {
	caKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate CA key: %w", err)
	}

	caParams := &x509.Certificate{
		SerialNumber: big.NewInt(1),
		Subject: pkix.Name{
			Organization: []string{"TestPKI Server Acceptor"},
			CommonName:   "TestPKI CA",
		},
		NotBefore:             time.Now(),
		NotAfter:              time.Now().Add(time.Hour * 24 * 365),
		IsCA:                  true,
		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
		MaxPathLen:            1,
		BasicConstraintsValid: true,
	}

	caBytes, err := x509.CreateCertificate(rand.Reader, caParams, caParams, &caKey.PublicKey, caKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create CA certificate: %w", err)
	}

	caPEM, err := EncodeX509Certificate(caBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to encode CA certificate: %w", err)
	}

	caCert, err := DecodeX509Certificate(caPEM)
	if err != nil {
		return nil, fmt.Errorf("failed to decode CA certificate: %w", err)
	}

	serverKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate server key: %w", err)
	}

	serverCSRBytes, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{
		SignatureAlgorithm: x509.ECDSAWithSHA256,
	}, serverKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create server CSR: %w", err)
	}

	serverCSR, err := DecodeX509CSR(serverCSRBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to decode server CSR: %w", err)
	}

	serverParams := &x509.Certificate{
		SerialNumber:       big.NewInt(1),
		Signature:          serverCSR.Signature,
		SignatureAlgorithm: x509.ECDSAWithSHA256,
		PublicKeyAlgorithm: x509.ECDSA,
		PublicKey:          serverCSR.PublicKey,
		Subject: pkix.Name{
			CommonName: "TestPKI Server",
		},
		DNSNames:    []string{"localhost"},
		NotBefore:   time.Now(),
		NotAfter:    time.Now().Add(time.Hour * 24 * 365),
		KeyUsage:    x509.KeyUsageDigitalSignature,
		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
	}

	serverBytes, err := x509.CreateCertificate(rand.Reader, serverParams, caCert, serverCSR.PublicKey, caKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create server certificate: %w", err)
	}

	serverPEM, err := EncodeX509Certificate(serverBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to encode server certificate: %w", err)
	}

	clientKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate client key: %w", err)
	}

	clientCSRBytes, err := x509.CreateCertificateRequest(rand.Reader, &x509.CertificateRequest{
		SignatureAlgorithm: x509.ECDSAWithSHA256,
	}, clientKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create client CSR: %w", err)
	}

	clientCSR, err := DecodeX509CSR(clientCSRBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to decode client CSR: %w", err)
	}

	clientParams := &x509.Certificate{
		SerialNumber:       big.NewInt(1),
		Signature:          clientCSR.Signature,
		SignatureAlgorithm: x509.ECDSAWithSHA256,
		PublicKeyAlgorithm: x509.ECDSA,
		PublicKey:          clientCSR.PublicKey,
		Subject: pkix.Name{
			CommonName: "TestPKI Client",
		},
		NotBefore:   time.Now(),
		NotAfter:    time.Now().Add(time.Hour * 24 * 365),
		KeyUsage:    x509.KeyUsageDigitalSignature,
		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
	}

	clientBytes, err := x509.CreateCertificate(rand.Reader, clientParams, caCert, clientCSR.PublicKey, caKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create client certificate: %w", err)
	}

	clientPEM, err := EncodeX509Certificate(clientBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to encode client certificate: %w", err)
	}

	caPool := x509.NewCertPool()
	caPool.AddCert(caCert)

	return &TestPKI{
		CaCert:     caPEM,
		ClientCert: clientPEM,
		ClientKey:  EncodeECDSAPrivateKey(clientKey),
		ClientConfig: &tls.Config{
			Certificates: []tls.Certificate{{Certificate: [][]byte{clientBytes}, PrivateKey: clientKey}},
			ServerName:   "localhost",
			RootCAs:      caPool,
		},
		ServerCert: serverPEM,
		ServerKey:  EncodeECDSAPrivateKey(serverKey),
		ServerConfig: &tls.Config{
			Certificates: []tls.Certificate{{Certificate: [][]byte{serverBytes}, PrivateKey: serverKey}},
			ClientCAs:    caPool,
			ClientAuth:   tls.RequireAndVerifyClientCert,
		},
	}, nil
}

func EncodeECDSAPrivateKey(privateKey *ecdsa.PrivateKey) []byte {
	marshalled, _ := x509.MarshalPKCS8PrivateKey(privateKey)
	return pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: marshalled})
}

func DecodeECDSAPrivateKey(encoded []byte) (*ecdsa.PrivateKey, error) {
	block, _ := pem.Decode(encoded)
	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	if privateKey, ok := key.(*ecdsa.PrivateKey); ok {
		return privateKey, nil
	}
	return nil, fmt.Errorf("failed to decode private key")
}

func EncodeX509Certificate(caBytes []byte) ([]byte, error) {
	caPEMBuffer := new(bytes.Buffer)
	err := pem.Encode(caPEMBuffer, &pem.Block{
		Type:  "CERTIFICATE",
		Bytes: caBytes,
	})
	if err != nil {
		return nil, err
	}

	return caPEMBuffer.Bytes(), nil
}

func DecodeX509Certificate(encoded []byte) (*x509.Certificate, error) {
	block, _ := pem.Decode(encoded)
	return x509.ParseCertificate(block.Bytes)
}

func DecodeX509CSR(encoded []byte) (*x509.CertificateRequest, error) {
	block, _ := pem.Decode(encoded)
	if block == nil {
		return x509.ParseCertificateRequest(encoded)
	} else {
		return x509.ParseCertificateRequest(block.Bytes)
	}
}
